service/s3/s3manager/upload.go

```diff
@@ -638,6 +706,43 @@ func (u *multiuploader) send(c chunk) error {
 	return nil
 }
 
+// isChunkUploaded is called before uploading each chunk to determine
+// if this chunk can be skipped. Chunks can be skipped when resuming an upload
+// if the ETag matches for the previously uploaded chunk.
+func (u *multiuploader) isChunkUploaded(c chunk) bool {
+	if !u.resumed {
+		return false
+	}
+
+	var expectedETag string
+
+	u.m.Lock()
+	for _, part := range u.parts {
```

Comment:
Is the lock needed here? If `u.parts` is only written to once before the uploads begin it should be safe to iterate through the parts concurrently. Since the parts list may be a sparse slice, a map may be a better data structure to use for these. Where the part number is the map key.

Reply:
`u.parts` is written and read concurrently by `multiuploader.send()`, so I think the lock is necessary.